### ISSUE TITLE
perf

### ISSUE THREAD

**Issue by ghost** [veonim/veonim#40]
is clearLine + fillText faster than getimagedata + putimagedata?

could keep internal buffer of screen chars and "scroll" region via fillText instead of copying all pixels.

-or-

create huge canvas buffer (overflowing), scroll-y the div instead. updates outside of visible buffer are delayed.
cons: vim has to render a shit ton

-or-

track "update buffer diff only" neovim issue and create own internal buffer representation... mehhbe

**Comment by goavn** [comment]
batch fillStyle calls

**Comment by goavn** [comment]
pre-render?

https://www.html5rocks.com/en/tutorials/canvas/performance/

can we split chunks of rendering to separate threads, render to prerendered imagedata then composite in main thread?

**Comment by goavn** [comment]
ctx.imageSmoothingEnabled = false

**Comment by goavn** [comment]
like OP:

keep grid1 + grid2 (textual arrays - typed arrays?) and diff to create more optimized/batched patches

**Comment by goavn** [comment]
overlaid canvases. swap them out lol. 1-2-1-2-1-2-1-2

**Comment by goavn** [comment]
smaller grids of canvases? 4x4 vs 1

**Comment by goavn** [comment]
clearRect > fillRect for clearing. if use bg + fg layers and composite, would it be faster? (we have to color bg to a color instead of transparent black)

**Comment by goavn** [comment]
make sure all cords are ints not floats

**Comment by goavn** [comment]
composite layers (alpha: true) + clearRect is slower than just alpha: false + fillStyle/fillRect

**Comment by goavn** [comment]
a significant increase if concat put chars together into string and fillText string vs string.forEach(char => fillText(char))

but alignment is off... are whitespaces getting truncated in trim()?

**Comment by goavn** [comment]
if bg color didn't change, do clearRect instead of fillRect(bg)

**Comment by goavn** [comment]
if whitespace, don't render char

**Comment by deomitrus** [comment]
drawImage from existing canvas to copy? have a second "buffer" canvas and drawImage between?

tl;dr (get|put)ImageData brings data to JS (slow). just want to move data as close to the metal as possible

**Comment by deomitrus** [comment]
find hax from `sourcelair/xterm.js/pull/938`
https://code.visualstudio.com/blogs/2017/10/03/terminal-renderer

**Comment by goavn** [comment]
https://medium.com/@weihanglo/source-reading-how-is-new-terminal-in-vs-code-so-fast-10a40f7f8792

**Comment by goavn** [comment]
render text to canvas before api.input()

**Comment by deomitrus** [comment]
damn son, canvas is as fast as the roadrunner

**Comment by deomitrus** [comment]
this issue is way too broad. i think this issue was about improving performance of canvas renderer. i would categorize this in two parts:

- improve scroll perf (move canvas region from one place to another)
- improve text drawing (redraw.put events, etc.)

first part is done. it's good and fast.
the second part will involve using font atlas/bitmaps.

the second issue is tracked here: #195 

closing this for now. thanks for watching. until next time.
